File: game/game.py
import time
import validation as validation
from core.board_state import BoardState
from core.move import Move
from core.action import ActionType


class Game:

    def __init__(self, players):
        self.players = players
        self.player_index = 0
        self.board = BoardState()
        self.reset()

    def play(self, print_boards=False):
        while self.winner is None:
            self.take_turn()

    def take_turn(self):
        if self.winner is not None:
            return
        player = self.players[self.player_index]
        opponent = self.players[player.opp_index]
        action = player.take_action(opponent, self.board)
        move = Move(self.board, player.index, action)
        is_move_valid = validation.validate_move(player, opponent, move)
        if not is_move_valid:
            self.winner = player.opp_index
            return
        self.moves.append(move)
        self.board = move.get_board_state()
        if move.action.type == ActionType.MOVE:
            player.pos = move.action.new_pos
        else:
            player.wall_count -= 1

        if player.pos.y == player.goal_row:
            self.winner = self.player_index + 1

        # Switch players
        self.player_index = opponent.index
    
    def reset(self):
        self.moves = []
        self.player_index = 0
        self.winner = None
        player1 = self.players[0]
        player2 = self.players[1]
        player1.reset()
        player2.reset()
        self.board = BoardState()
        self.board.player_positions = []
        self.board.player_positions.append(player1.pos)
        self.board.player_positions.append(player2.pos)
        self.board.player_walls = [10, 10]

File: game/src/game.cc
#include "game.h"
#include "validation.h"
#include <iostream>

using namespace std;

Game::Game(Player* player_1, Player* player_2, bool validate_actions, bool store_turns, const BoardState& initial_board) :
    validate_actions_(validate_actions),
    store_turns_(store_turns) {
    players_ = array<Player*, 2>();
    players_[0] = player_1;
    players_[1] = player_2;
    player_1->SetIndex(0);
    player_2->SetIndex(1);
    Reset(initial_board);
}

void Game::Play() {
    while (!TakeTurn()) {        
    }
}

bool Game::TakeTurn() {
    if (winning_player_index_ != -1) {
        return true;
    }

    auto current_player = players_[current_player_index_];
    const auto action = current_player->TakeAction(current_board_);
    
    if (validate_actions_ && !ValidateAction(current_board_, current_player_index_, action, true)) {
        cout << "Player " << current_player_index_ << " Made An Invalid Move! (" << action << ")" << endl;
        return true;
    }
    
    current_board_ = BoardState(current_board_, action, current_player_index_);
		
    if (store_turns_)
        turns_.emplace_back(current_board_, current_player_index_, action);
    turn_count_ += 1;

    if (current_board_.GetPlayerDistance(current_player_index_) == 0) {
        winning_player_index_ = current_player_index_;
        return true;
    }

    current_player_index_ = 1 - current_player_index_;
    return false;
}

void Game::Reset(const BoardState& initial_board) {
    if (store_turns_) {
	    turns_.clear();
    }
    turn_count_ = 0;
    current_player_index_ = 0;
    winning_player_index_ = -1;
    initial_board_ = initial_board;
    current_board_ = initial_board;
}

int Game::GetWinner() const {
    return winning_player_index_;
}

Turn Game::GetTurn(const int index) const {
    return turns_[index];
}

int Game::GetTurnCount() const {
    return turn_count_;
}

const BoardState& Game::GetBoardState() const {
    return current_board_;
}

const BoardState& Game::GetInitialBoardState() const {
    return initial_board_;
}
